docs/snode.rst

```diff
@@ -44,33 +46,69 @@ See :ref:`layout` for more details. ``ti.root`` is the root node of the data str
         x.shape() # returns (3, 5, 4)
 
 
-.. function:: snode.get_shape(index)
+.. function:: tensor.dim()
+
+    :parameter tensor: (Tensor)
+    :return: (scalar) the dimensionality of the tensor
+
+    Equivalent to ``len(tensor.shape())``.
+
+    ::
+
+        ti.root.dense(ti.ijk, (8, 9, 10)).place(x)
+        x.dim()  # 3
+
+
+.. function:: tensor.snode()
+
+    :parameter tensor: (Tensor)
+    :return: (SNode) the structual node where ``tensor`` is placed
+
+    ::
+
+        x = ti.var(dt=ti.i32)
+        y = ti.var(dt=ti.f32)
+        ti.root.place(x, y)
+        x.snode()
+
+
+.. function:: snode.shape()
 
     :parameter snode: (SNode)
     :parameter index: axis (0 for ``i`` and 1 for ``j``)
     :return: (scalar) the size of tensor along that axis
 
-    Equivalent to ``tensor.shape()[i]``.
+    Equivalent to ``tensor.shape()``.
 
     ::
 
-        ti.root.dense(ti.ijk, (3, 5, 4)).place(x)
-        x.snode().get_shape(0)  # 3
-        x.snode().get_shape(1)  # 5
-        x.snode().get_shape(2)  # 4
+        blk1 = ti.root
+        blk2 = blk1.dense(ti.i,  3)
+        blk3 = blk2.dense(ti.jk, (5, 2))
+        blk4 = blk3.dense(ti.k,  2)
+        blk1.shape()  # ()
+        blk2.shape()  # (3, )
+        blk3.shape()  # (3, 5, 2)
+        blk4.shape()  # (3, 5, 4)
 
 
-.. function:: tensor.dim()
+.. function:: snode.get_shape(index)
 
-    :parameter tensor: (Tensor)
-    :return: (scalar) the dimensionality of the tensor
+    :parameter snode: (SNode)
+    :parameter index: axis (0 for ``i`` and 1 for ``j``)
+    :return: (scalar) the size of tensor along that axis
 
-    Equivalent to ``len(tensor.shape())``.
+    Equivalent to ``snode.shape()[i]``, used in Taichi-scope to prevent `[...]` being transformed to tensor access.
```

Comment:
IIRC `snode.shape` gives you a tuple and using `[]` on that doesn't get transformed to tensor access?

Reply:
Seems we treated specially in `ti.subscript`? Cool!
```
Before preprocessing:
@ti.kernel
def func():
    print(x.snode().shape()[1])

After preprocessing:
def func():
  ti.ti_print(ti.subscript(x.snode().shape(), 1))
```

Reply:
So can we deprecate `get_shape(i)` and use `shape()[i]` now according to the python exactly-one-way-to-do-one-thing rule?

Reply:
Can't agree more, many users are confusing `dim` with vector dimensions. Things we could do in another PR:
- [ ] deprecate `dim()`.
- [ ] make `shape()` a `@property`.
- [ ] document `mat.n`, `mat.m` in `meta.rst`.